src/crimson/os/seastore/lba_mapping.h

```diff
@@ -4,69 +4,208 @@
 #pragma once
 
 #include "crimson/os/seastore/cached_extent.h"
-#include "crimson/os/seastore/btree/btree_range_pin.h"
+#include "crimson/os/seastore/btree/btree_types.h"
 #include "crimson/os/seastore/lba_manager/btree/lba_btree_node.h"
 #include "crimson/os/seastore/logical_child_node.h"
 
 namespace crimson::os::seastore {
 
-class LBAMapping;
-using LBAMappingRef = std::unique_ptr<LBAMapping>;
+namespace lba_manager::btree {
+class BtreeLBAManager;
+}
 
-class LogicalCachedExtent;
-
-class LBAMapping : public BtreeNodeMapping<laddr_t, paddr_t> {
+class LBAMapping {
 public:
-  LBAMapping(op_context_t<laddr_t> ctx)
-    : BtreeNodeMapping<laddr_t, paddr_t>(ctx) {}
-  template <typename... T>
-  LBAMapping(T&&... t)
-    : BtreeNodeMapping<laddr_t, paddr_t>(std::forward<T>(t)...)
-  {}
-
-  // An lba pin may be indirect, see comments in lba_manager/btree/btree_lba_manager.h
-  virtual bool is_indirect() const = 0;
-  virtual laddr_t get_intermediate_key() const = 0;
-  virtual laddr_t get_intermediate_base() const = 0;
-  virtual extent_len_t get_intermediate_length() const = 0;
-  // The start offset of the pin, must be 0 if the pin is not indirect
-  virtual extent_len_t get_intermediate_offset() const = 0;
-
-  virtual get_child_ret_t<lba_manager::btree::LBALeafNode, LogicalChildNode>
-  get_logical_extent(Transaction &t) = 0;
-
-  void link_child(LogicalChildNode *c) {
-    ceph_assert(child_pos);
-    child_pos->link_child(c);
-  }
-  virtual LBAMappingRef refresh_with_pending_parent() = 0;
+  LBAMapping() = default;
+  LBAMapping(LBACursorRef direct, LBACursorRef indirect)
+    : direct_cursor(std::move(direct)),
+      indirect_cursor(std::move(indirect))
+  {
+    assert(!is_null());
```

Comment:
Following the above comments, can we make sure a LBAMapping can only be either null or is_linked_direct ?

Specifically, replace `assert(!is_null())` by `assert(is_linked_direct())` completely.

Reply:
I think we could drop the default ctor that is the only approach to construct a null LBAMapping.

Related changes: make `mapping` field of `BtreeLBAManager::update_refcount_ret_bare_t` to `std::optional`